src/js/structs/JobTree.js

```diff
@@ -0,0 +1,110 @@
+import Job from './Job';
+import Tree from './Tree';
+
+module.exports = class JobTree extends Tree {
+  /**
+   * (Chronos) JobTree
+   * @param {{
+   *          id:string,
+   *          items:array<({id:string, items:array}|*)>,
+   *          filterProperties:{propertyName:(null|string|function)}
+   *        }} options
+   * TODO: DCOS-7369 - Add remove item method
+   * @constructor
+   * @struct
+   */
+  constructor({id, items, filterProperties} = {}) {
+    super({filterProperties});
+
+    this.id = '/';
+
+    if (typeof id == 'string') {
+      this.id = id;
+    }
+
+    // Parse and add items to the current tree
+    if (Array.isArray(items)) {
+      items.forEach((item)=> {
+        // Create instances of JobTree for tree like items and add them
+        if ((item.items != null && Util.isArray(item.items))
+          && !(item instanceof JobTree)) {
+          this.add(new this.constructor(
+            Object.assign({filterProperties: this.getFilterProperties()}, item)
+          ));
+
+          return;
+        }
+
+        // Convert items into instance of Job and add them to the current tree
+        if (!(item instanceof Job) && !(item instanceof JobTree)) {
+          this.add(new Job(item));
+
+          return;
+        }
+
+        this.add(item);
+      })
+    }
+  }
+
+  /**
+   * Add item - This method will create sub trees if needed to insert the item
+   * at the correct location (based on id/path matching).
+   * @param {JobTree|Job} item
+   */
+  add(item) {
+    if (!(item instanceof Job || item instanceof JobTree)) {
+      throw new TypeError('item is neither an instance of Job nor JobTree');
+    }
+
+    if (!item.getId().startsWith(this.getId())) {
+      throw new Error(
+        `item id (${item.getId()}) doesn't match tree id (${this.getId()})`
+      );
+    }
+
+    // Check if the item is already present in the tree (including sub trees)
+    if (this.findItemById(item.getId())) {
+      // TODO: DCOS-7370 - Update item (type Job) data
+      return;
+    }
+
+    // Get the parent id (e.g. /group) by matching every thing but the item
+    // name including the preceding slash "/" (e.g. /id).
+    let [parentId] = item.getId().match(/\/.*?(?=\/?[^/]+\/?$)/);
+
+    // Add item to the current tree if it's the actual parent tree
```

Comment:
We're trying to find the parent of an item (`Job` or `JobTree`), by extracting the`parentId` from the `itemId`. We're adding it to the current tree if the `parentId` matches the current `JobTree` id, otherwise we're trying to find the parent (`findItemById`) or create a new parent (`JobTree`) and add the item to that one.
**TLDR;** This method recursively creates all the parents needed to insert an item at the correct spot (based on it's id).